### PR TITLE
Fix/collection output transpile imports

### PR DESCRIPTION
**Pull request by tanner-reits**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [X] Build (`npm run build`) was run locally and any changes were pushed
- [X] Unit tests (`npm test`) were run locally and passed
- [X] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [X] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [X] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

The `dist-collection` output target currently does not transpile path aliases defined in a project's `tsconfig.json`. Instead, the import paths are left unchanged and thus reference modules that cannot be resolved.

GitHub Issue Number: https://github.com/ionic-team/stencil/issues/2815

## What is the new behavior?

- A transformation step has been added as a part of the `dist-collection` output target that will transpile path aliases and convert them in to relative paths based on the output files from the TS compilation

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

## Testing

Unit tests have been added for the new transformer, and existing unit tests continue to pass. (Further testing will be possible once infrastructure is in place related to STENCIL-223).

E2E tests continue to pass (as consistently as they did before)

Validated output on test repos (including Ionic framework)

### CODE REVIEW

File: src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts
Comment (tanner-reits) [inline src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts:20]: Failing test. The testing helpers we have do not account for the need to test with multiple JS modules.

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [inline src/compiler/transformers/map-imports-to-path-aliases.ts:25]: This is what is causing tests to fail for this transformer. In the testing context, TS is not able to resolve the import modules, so import paths are always left untouched.

File: src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts
Comment (rwaskiewicz) [reply src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts:20]: So, this is tricky. We need to restore some functionality where we have a compiler instance along with a virtual filesystem of some kind, and a patched TS compiler (all for testing things in a more 'e2e' fashion).  In the mean time, for unit testing, does spying on this TS method make things go away? TAL at the patch I attached to this comment (GH doesn't support '.patch' extensions, go figure 🙄). But you should be able to apply it with `git apply` all the same.

[ts_spy.txt](https://github.com/ionic-team/stencil/files/9338991/ts_spy.txt)

I didn't take a super close look here just yet, but this should unblock you at the very least on the testing front

Comment (tanner-reits) [reply src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts:20]: @rwaskiewicz I can do something like that, but at that point I question if that test really adds any benefit since it's not actually performing the action we're relying on TS for 🤷‍♂️ The test doesn't blow up, it just fails cuz the text doesn't get replaced like it does in a build.

With that said, we won't get "correct" test coverage either way until STENCIL-223 is fixed. So, I'm curious how we wanna handle this. Feels like we'd either need a dev build to have this verified by users (and framework team) or some kind of config flag. I know we had considered this being a potential breaking change too

Comment (tanner-reits) [reply src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts:20]: Fair points! I'll get those tests and some general code comments added in another commit. I'll also add a "todo" to get the actual module resolution step tested once STENCIL-223 is done

File: src/compiler/output-targets/dist-collection/index.ts
Comment (tanner-reits) [inline src/compiler/output-targets/dist-collection/index.ts:38]: Only downside to this approach is, for this target, we'll essentially end up transpiling each module twice. However, I ran a couple builds with this change and without and only seemed to affect Stencil build times within a project by a second or so (that testing was done with the framework code base)

**Comment by tanner-reits** [comment]
Last thing we should need to decide is how to handle this release/deployment. Basically if we feel like this can be merged directly, considered a breaking change for V3, or something we would want to keep behind a config flag.

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (rwaskiewicz) [inline src/compiler/transformers/map-imports-to-path-aliases.ts:16]: Since this is only called by `src/compiler/output-targets/dist-collection/index.ts#outputCollection`, and `outputCollection` passes a config of type `ValidatedConfig` to this fn, can we update this function signature accordingly?
```suggestion
export const mapImportsToPathAliases = (config: d.ValidatedConfig): ts.TransformerFactory<ts.SourceFile> => {
```

File: src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts
Comment (rwaskiewicz) [inline src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts:25]: Consider making these test descriptions assertive to minimize repetition of 'should':
```suggestion
  it('ignores relative imports', () => {
```

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
> Last thing we should need to decide is how to handle this release/deployment. Basically if we feel like this can be merged directly, considered a breaking change for V3, or something we would want to keep behind a config flag.

I think we should add a configuration flag to the `collections` output target for folks to try this out and report any issues to us. My thoughts are we do the following:
- resolve comments from @alicewriteswrongs and I on this PR
- Create a separate PR that targets this one that adds the flag
  - We'll also want to update our internal doc for flags for V3 
- Create a sibling PR with the associated documentation

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [reply src/compiler/transformers/map-imports-to-path-aliases.ts]: Ah yep, good catch. Couldn't use the ValidatedConfig when this was running as a part of the main transpilation process

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [reply src/compiler/transformers/map-imports-to-path-aliases.ts:24]: Yes this is needed. Using the approach you suggested errors out. `node.getSourceFile()` returns `undefined`

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [reply src/compiler/transformers/map-imports-to-path-aliases.ts]: Good catch. I changed this to replace all file extensions defined on `ts.Extensions`

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (rwaskiewicz) [inline src/compiler/transformers/map-imports-to-path-aliases.ts:51]: I think this is close, but still not quite right.  That's on me, I think my previous comment/concern wasn't worded well. In the initial implementation:

```ts
relative(sourceFile, module.resolvedModule.resolvedFileName).replace(module.resolvedModule.extension, '')
```

we would always replace the first instance of anything that matched the resolved module's extension. So if we had something like
```
foo.mjs.mjs
```
this would become
```
foo.mjs
```
which seemed correct, until we grouped the extensions:
```
foo[.mjs][.mjs]
// the first '.mjs' is what would get removed
foo[.mjs] // <- second extension stuck around
```


I think we still run into an issue here, as we might be removing more than we'd like for cases like
```
foo.mjs.js
```
we'd end up removing all the extensions, which I don't _think_ is what we want. Instead, we should rm the last entry only with something like:

```diff
- const extensionRegex = new RegExp(Object.values(ts.Extension).join('|'), 'g');
+ const extensionRegex = new RegExp(Object.values(ts.Extension).map((ext) => `${ext}$`).join('|'));
```

where:
- we append '$' to each extension to check that the filename ends with the extension
- we remove the 'g' flag, since we only want to remove one thing

Does that make sense? Or did I miss something here? 😆

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (rwaskiewicz) [inline src/compiler/transformers/map-imports-to-path-aliases.ts:21]: Can we add a few test cases for these, where
- `destinationFilePath` is a parent of `destinationDirectory` and both are non-empty strings?
- `destinationDirectory` is a parent of `destinationFilePath` and both are non-empty strings?
- both are the same directory and both are non-empty strings?

this might get a little funky, since we'll probably have to rely on `normalize` in our tests for Win/Unix compatability

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
Thanks! Just a few more comments on the revisions

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [reply src/compiler/transformers/map-imports-to-path-aliases.ts:51]: @rwaskiewicz This was kinda where I started, but then I tested a file with weird extensions like `foo.mts.ts`. This resulted in an import path like `../foo.mts` which throws an error because it's not a JS file import. Not sure how realistic files like that are, but it was causing issues

Comment (rwaskiewicz) [reply src/compiler/transformers/map-imports-to-path-aliases.ts:51]: Yeah, IDK of realistic it is, kind of a weird edge case for someone to have something like that working in Stencil 🤷

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (alicewriteswrongs) [reply src/compiler/transformers/map-imports-to-path-aliases.ts:24]: it looks to me like you have it up here because you can't get the right value until the function returned here is called:

https://github.com/ionic-team/stencil/blob/88a6974882c23e97fbd257e81cd8b1349235345e/src/compiler/transformers/map-imports-to-path-aliases.ts#L83-L79

and then you need to use it within the `visit` body. I think doing `let foo: string;` does always create a `strictNullChecks` error though because since it's not initialized with a string the variable is `undefined` (and therefore not a string).

One alternative approach could be to curry the `sourceFile` argument like this:

```ts
const visit = (sourceFile: string) => (node: ts.Node): ts.VisitResult<ts.Node> => {
```

then below you could do something like

```ts
    return (tsSourceFile) => {
      const sourceFile = tsSourceFile.fileName;

      return ts.visitEachChild(tsSourceFile, visit(sourceFile), transformCtx);
    };
```

which would I think fix the `strictNullChecks` error and is arguably "more correct" code since we don't declare a variable without initializing it and then access it later w/ the implicit assumption it's been assigned in the intervening time

**Review by alicewriteswrongs — CHANGES_REQUESTED** [review]
two thoughts / questions

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [reply src/compiler/transformers/map-imports-to-path-aliases.ts:51]: @rwaskiewicz with that said, do we feel comfortable removing just the last extension, or want to play it safe and remove all?

Comment (rwaskiewicz) [reply src/compiler/transformers/map-imports-to-path-aliases.ts:51]: I say remove the last extension. If someone reports a bug, they can directly link this comment and blame me lol 😆

File: src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts
Comment (tanner-reits) [reply src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts]: This was a good catch. I didn't actually need to pull in that config anymore with how we handled these tests so I just stripped that entirely.

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [reply src/compiler/transformers/map-imports-to-path-aliases.ts]: This worked! Thanks for the suggestion!

File: src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts
Comment (alicewriteswrongs) [reply src/compiler/transformers/test/map-imports-to-path-aliases.spec.ts]: so right now basically the TS stuff is all being mocked out anyhow right? I didn't quite realize that when I first read through the tests

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (rwaskiewicz) [inline src/compiler/transformers/map-imports-to-path-aliases.ts:51]: Small bug - `join` will only be applied when there are >1 entries, and when it is applied, won't be applied to the last entry. So if `ts.Extension` was `['.js', '.ts', '.mjs']`, this only gives us:
```
new RegExp(Object.values(['.js', '.ts', '.mjs']).join('$|')) // Yields  /.js$|.ts$|.mjs/
```
I think we want something more like:
```
new RegExp(Object.values(['.js', '.ts', '.mjs']).map((ext) => `${ext}$`).join('|')) // Yields /.js$|.ts$|.mjs$/
```

Specifically:
```
new RegExp(Object.values(ts.Extension).map((ext) => `${ext}$`).join('|'))
```

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

File: src/compiler/transformers/map-imports-to-path-aliases.ts
Comment (tanner-reits) [reply src/compiler/transformers/map-imports-to-path-aliases.ts]: Oof yeah you're right, silly mistake on my part. Thanks for the callout!

**Comment by rwaskiewicz** [comment]
LGTM 👍 Gonna hold off on the official approval until we decide how we want to handle the child PR(s)

**Review by alicewriteswrongs — APPROVED** [review]
I think this looks good! 👍

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**
